Code Coverage for integration tests scenarios

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#220 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have ran the wallet server in a separate thread. This way, we prevent the scenario from escaping code coverage :) 

# Comments

<!-- Additional comments or screenshots to attach if any -->

Note: I've also removed the 'Local' network target as it's not really a thing and isn't much use anyway.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented:
@akegalj Any hope with that one :) ?

akegalj commented on src/Cardano/Wallet/Kernel/Transactions.hs:
```diff
@@ -594,6 +596,27 @@ instance Arbitrary SignTransactionError where
         ]
 
+-- Which address scheme is being used
+data AddressSchemeVersion
```
special type created to help distinguish between old and new derivation scheme

akegalj commented on stack.yaml:
```diff
@@ -4,7 +4,7 @@ packages:
 extra-deps:
 - git: https://github.com/input-output-hk/cardano-sl
-  commit: cdc77284bde949d4b1b9ac7be81b76b906a71fc7
+  commit: 4ecd69bebd98c27d1d20d68c64293eb303a6cf8c
```
partially extended key generation with new method https://github.com/input-output-hk/cardano-sl/commit/4ecd69bebd98c27d1d20d68c64293eb303a6cf8c

akegalj replied:
Ok - I will start moving the logic to cardano-wallet within this PR

akegalj commented on test/unit/Test/Spec/Ed25519Bip44.hs:
```diff
@@ -23,6 +23,10 @@ import           Test.QuickCheck (Arbitrary (..), InfiniteList (..), Property,
                      arbitraryBoundedIntegral, arbitrarySizedBoundedIntegral,
                      property, shrinkIntegral, (.&&.), (===), (==>))
 
+-- TODO (akegalj): seems like there already is a notion of
```
there is a development point about this

akegalj commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -72,11 +75,12 @@ import           Servant.Server
+    , fixtureAccountId  :: HdAccountId
+    , fixturePw         :: PassiveWallet
+    , fixturePayees     :: NonEmpty (Address, Coin)
+    , fixturePassPhrase :: PassPhrase
```
passphrase is required in ed25519bip44

akegalj commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -87,39 +91,54 @@ prepareFixtures :: NetworkMagic
                 -> Fixture.GenActiveWalletFixture Fixture
 prepareFixtures nm initialBalance toPay = do
-    let (_, esk) = safeDeterministicKeyGen (B.pack $ replicate 32 0x42) mempty
+    addressVersion <- return Kernel.AddressSchemeV1 -- pick arbitrary
```
hardcoded in order to test V0 branch first and then V1

akegalj commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -87,39 +91,54 @@ prepareFixtures :: NetworkMagic
 prepareFixtures nm initialBalance toPay = do
-    let (_, esk) = safeDeterministicKeyGen (B.pack $ replicate 32 0x42) mempty
+    addressVersion <- return Kernel.AddressSchemeV1 -- pick arbitrary
+    passPhrase <- pick $ arbitrary @PassPhrase
```
passphrase is required for ed25519bip44

akegalj commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -87,39 +91,54 @@ prepareFixtures :: NetworkMagic
+    mnemonic <- pick $ arbitrary @(Mnemonic 12)
+    let seed = mnemonicToSeed mnemonic
+    let esk =
+            case addressVersion of
```
root key depends on derivation scheme

akegalj commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -87,39 +91,54 @@ prepareFixtures :: NetworkMagic
-                                                               esk
-                                                               (newAccountId ^. hdAccountIdIx . to getHdAccountIx)
-                                                               (getHdAddressIx newIndex)
+                        newIndex <- deriveIndex (pick . choose) HdAddressIx SoftDerivation
```
ed25519bip44 requires address derivation to be soft ferivation (non-hardened)

KtorZ replied:
That is a bit wrong here. Because the point of the new derivation is to be sequential, not random. So, it's likely a bit more complex than that. We need to generate addresses within a certain gap if we want the fixture to be meaningful. Otherwise, we won't be able to discover those addresses.

akegalj replied:
Yes - that's true.

I thought that generating in random won't harm for tests - but in fact it might - if rest of the code expects it to be sequential.

akegalj replied:
This commit https://github.com/input-output-hk/cardano-wallet/pull/223/commits/cf95ddb11a234c3f54e403d70c0f215aae23b9e5 should address this problem: creates simple sequential account and corresponding sequential addresses

akegalj replied:
Its a simplest use case and can be improved upon though

akegalj commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -87,39 +91,54 @@ prepareFixtures :: NetworkMagic
+                        newIndex <- deriveIndex (pick . choose) HdAddressIx SoftDerivation
+
+                        let Just (addr, _) =
+                                case addressVersion of
```
depending on derivation scheme picked - we are going to use old derivation method or ed25519bip44 (new one)

KtorZ commented on src/Cardano/Wallet/Kernel/Transactions.hs:
```diff
@@ -594,6 +596,27 @@ instance Arbitrary SignTransactionError where
+    = AddressSchemeV0
+    -- Address scheme 1 is bip44 HD wallet derived
+    -- address. It is derived with ed25519v1 curve and contains no address payload
+    | AddressSchemeV1
```
I'd however desambiguate the names here and go for `AddressSchemeRnd` and `AddressSchemeSeq`.

akegalj replied:
Done in https://github.com/input-output-hk/cardano-wallet/pull/223/commits/47f521beb310e23ef8a4739e12503d9a33a7189f

KtorZ commented on test/unit/Test/Spec/NewPayment.hs:
```diff
@@ -87,39 +91,54 @@ prepareFixtures :: NetworkMagic
-            hdAccountId = Kernel.defaultHdAccountId newRootId
-            hdAddress   = Kernel.defaultHdAddress nm esk emptyPassphrase newRootId
+            hdAccountId = newAccountId -- Kernel.defaultHdAccountId newRootId
+            hdAddress   = Nothing -- Kernel.defaultHdAddress nm esk passPhrase newRootId
```
To the best of my knowledge, the default address is there to allow users to perform transactions while the wallet is restoring (and therefore, hasn't discovered its addresses yet).
I am not sure how this is relevant in the context of unit testing...

KtorZ commented:
Closing this one until we are ready for this to get in.

KtorZ committed:
remove 'Local' Network target

KtorZ committed:
run the wallet server in the same thread as the test scenario

Before, we ran this in a separate process which, as a major downside, has been escaping any
form of coverage measurement. Having it ran in another _thread_ solves this problem

piotr-iohk reviewed (approved):
lgtm :+1:  (as there are some discussions around `cardano-wallet-launcher`, it can be done in another PR if needed)

piotr-iohk commented on lib/http-bridge/test/integration/Cardano/LauncherSpec.hs:
```diff
@@ -0,0 +1,36 @@
+                cancel handle
+            Right _ ->
+                expectationFailure
+                    "cardano-wallet-launcher isn't supposed to terminates. \
```
ah, typo (`terminates`)
